Spark on Kubernetes - basic scheduler backend

foxish opened the pull request:
Continuing https://github.com/apache-spark-on-k8s/spark/pull/492

Stripped out a lot of extraneous things, to create this. Our first PR upstream will likely be similar to this. (note that it is created against the master branch which is up-to-date.)

Following PRs will have:
- Submission Client
- Dynamic allocation
- HDFS locality fixes
etc

TODO before we can retarget this to apache/spark:master:
- [x] Renaming package to k8s (https://github.com/apache-spark-on-k8s/spark/pull/497)
- [x] Unit testing for executorpodfactory (https://github.com/apache-spark-on-k8s/spark/pull/491)

cc @ash211 @mccheah @apache-spark-on-k8s/contributors

foxish committed:
Spark on Kubernetes - basic scheduler backend

mccheah commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/ConfigurationUtils.scala:
```diff
@@ -0,0 +1,56 @@
+import org.apache.spark.{SparkConf, SparkException}
+import org.apache.spark.internal.Logging
+
+object ConfigurationUtils extends Logging {
```
Mark with `private[spark]`

mccheah commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/ConfigurationUtils.scala:
```diff
@@ -0,0 +1,56 @@
+import org.apache.spark.internal.Logging
+
+object ConfigurationUtils extends Logging {
+  def parseKeyValuePairs(
```
Where do we use this at least in the context of this PR?

mccheah commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala:
```diff
@@ -0,0 +1,466 @@
+      requestExecutorsService)
+
+  private val driverPod = try {
+    kubernetesClient.pods().inNamespace(kubernetesNamespace).
```
Might prefer to stylize a bit differently:

```
kubernetesClient.pods()
  .inNamespace(kubernetesNamespace)
  .withName(kubernetesDriverPodName)
  .get
```

mccheah commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala:
```diff
@@ -0,0 +1,466 @@
+      // by the executor pod watcher. If the loss reason was discovered by the watcher,
+      // inform the parent class with removeExecutor.
+      val disconnectedPodsByExecutorIdPendingRemovalCopy =
+          Map.empty ++ disconnectedPodsByExecutorIdPendingRemoval
```
I think there might be a more idiomatic way to copy here. Can you try `disconnectedPodsByExecutorIdPendingRemoval.toMap` - as an experiment, explicitly set the type of `disconnectedPodsByExecutorIdPendingRemovalCopy` to `Map[String, Pod]`? Basically as long as this collection is not mutable then we'll be safe.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactorySuite.scala:
```diff
@@ -0,0 +1,136 @@
+import org.apache.spark.deploy.kubernetes.constants
+import org.apache.spark.network.netty.SparkTransportConf
+
+class ExecutorPodFactoryImplSuite extends SparkFunSuite with BeforeAndAfter {
```
Nit - `ExecutorPodFactorySuite` - our convention at least so far has been to name the test after the trait and not the impl since there's only ever one impl for each trait. But this might change upstream.

mccheah reviewed (commented):
I think this is excellent content-wise for an initial push. We could have considered moving the executor failure handling to its separate commit, but I'm not too strongly opinionated one way or another.

mccheah commented:
Checkstyle on Java is what's causing the build failures.

kimoonkim commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala:
```diff
@@ -0,0 +1,466 @@
+  private val EXECUTOR_PODS_BY_IPS_LOCK = new Object
+  // Indexed by executor IP addrs and guarded by EXECUTOR_PODS_BY_IPS_LOCK
+  private val executorPodsByIPs = new mutable.HashMap[String, Pod]
+  private val podsWithKnownExitReasons: concurrent.Map[String, ExecutorExited] =
```
@foxish @varunkatta This part seems a bit outdated compared with branch-2.2-kubernetes, which use ConcurrentHashMap for this line.

kimoonkim reviewed (commented):
Compared this with `branch-2.2-kubernetes`. Looks reasonable to me.

Just one minor comment below. PTAL.

foxish commented:
Addressed comments. Thanks @mccheah and @kimoonkim!

foxish commented:
Unit tests look good. There's not much else that can be tested at this point. If there are no further comments, I'll turn this into a PR against upstream tomorrow at 5pm.

felixcheung reviewed (commented):
sounds good.
btw, is there a reason for the `private[spark] ` scope? I think it would be better to start with a more constrained scope, like `deploy` or `k8s` even. but we could see if anyone complains :)

mccheah commented:
@felixcheung `private[spark]` has been canonically used everywhere but I don't know if there is a set convention to scope these down further.

mccheah commented:
@foxish cherry-picking #491 for this too, correct?

foxish commented:
The relevant parts of #491 are already covered in this PR.

mccheah reviewed (approved):
So I see - this is good to go.

mccheah commented:
...Once the build is fixed, of course =)

felixcheung commented:
It should, for example sql has private[sql] and mesos has private[mesos]

https://github.com/apache/spark/blob/master/resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/MesosClusterDispatcher.scala

foxish committed:
Adding to modules.py and SparkBuild.scala

foxish committed:
Exclude from unidoc, update travis

foxish committed:
Address a bunch of style and other comments

foxish committed:
Fix some style concerns

tnachen commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/SparkKubernetesClientFactory.scala:
```diff
@@ -0,0 +1,103 @@
+ * options for different components.
+ */
+private[spark] object SparkKubernetesClientFactory {
+
```
nit: Clean up these white spaces, or at least be consistent

foxish commented:
This is being reviewed in https://github.com/apache/spark/pull/19468

foxish committed:
Clean up YARN constants, unit test updates

foxish committed:
Couple of more style comments

mccheah committed:
Address CR comments.

mccheah committed:
Extract initial executor count to utils class

mccheah committed:
Fix scalastyle

mccheah committed:
Fix more scalastyle

mccheah committed:
Pin down app ID in tests. Fix test style.

mccheah committed:
Address comments.

mccheah committed:
Various fixes to the scheduler

- Move Kubernetes client calls out of synchronized blocks to prevent
locking with HTTP connection lag
- Fix a bug where pods that fail to launch through the APi are not
retried
- Remove the map from executor pod name to executor ID by using the
Pod's labels to get the same information without having to track extra
state.

mccheah committed:
Address comments

foxish committed:
Update fabric8 client version to 3.0.0

liyinan926 committed:
Addressed more comments

liyinan926 committed:
One more round of comments

liyinan926 committed:
Added a comment regarding how failed executor pods are handled

liyinan926 committed:
Addressed more comments

liyinan926 committed:
Fixed Scala style error

liyinan926 committed:
Removed unused parameter in parsePrefixedKeyValuePairs

liyinan926 committed:
Another round of comments

liyinan926 committed:
Addressed latest comments

liyinan926 committed:
Addressed comments around licensing on new dependencies

liyinan926 committed:
Fixed unit tests and made maximum executor lost reason checks configurable

liyinan926 committed:
Removed default value for executor Docker image

liyinan926 committed:
Close the executor pod watcher before deleting the executor pods

liyinan926 committed:
Addressed more comments

felixcheung commented:
could you check the Integration Test? it seems to have failed to build `[ERROR] Could not find the selected project in the reactor: resource-managers/kubernetes/integration-tests @ `

mccheah commented:
@felixcheung we shouldn't be running integration tests when we submit upstream. Also, shouldn't we instead be tracking https://github.com/apache/spark/pull/19717 and thus this PR be closed?

foxish commented:
This PR is tracking the changes in the upstream one. It's the same branch
used for both PRs, and is simply for tracking.

On Nov 28, 2017 10:32 AM, "mccheah" <notifications@github.com> wrote:

> @felixcheung <https://github.com/felixcheung> we shouldn't be running
> integration tests when we submit upstream. Also, shouldn't we instead be
> tracking apache#19717 <https://github.com/apache/spark/pull/19717> and
> thus this PR be closed?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache-spark-on-k8s/spark/pull/498#issuecomment-347619545>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA3U5xdA-jTh_3QQGwd4KGQ-1C15LdWyks5s7FGxgaJpZM4PdiJJ>
> .
>

liyinan926 commented:
Closing this one as the upstream PR has been merged.
